<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [4]:
import pandas as pd
import psycopg2

import warnings
warnings.filterwarnings("ignore")


In [5]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [7]:
# текст запроса
query_3_1 = f'''
    SELECT
        COUNT(DISTINCT id)
    FROM public.vacancies
'''

In [8]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query_3_1, connection)

# Отображаем полученный результат
display(df)

count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [9]:
# SQL-запрос для подсчета количества уникальных значений:
# Подсчитываем количество уникальных значений в столбце id из таблицы public.employers
query = """
    SELECT
        COUNT(DISTINCT id)
    FROM public.employers
"""

In [10]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [11]:
# SQL-запрос для подсчета количества уникальных значений:
# Подсчитываем количество уникальных значений в столбце id из таблицы public.areas
query = f"""
    SELECT
        COUNT(DISTINCT id)
    FROM public.areas
"""

In [12]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [13]:
# SQL-запрос для подсчета количества уникальных значений:
# Подсчитываем количество уникальных значений в столбце id из таблицы public.industries
query = f"""
    SELECT
        COUNT(DISTINCT id)
    FROM public.industries
"""

In [14]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

count
0    294

***

# Предварительный анализ данных
На основании предоставленных запросов и результатов проведён более детальный анализ текущей базы данных.

---

## **1. Количество вакансий в базе**
- **Запрос:** Подсчет всех уникальных вакансий.
- **Результат:** В базе данных насчитывается **49197 вакансий**.
- **Вывод:**
  - Впечатляющее количество вакансий свидетельствует о значительном охвате рынка труда.
  - Важно учитывать, что не все вакансии могут быть актуальными: необходимо дальнейшее исследование для фильтрации устаревших и дублированных записей.
  - Это открывает возможности для анализа, какие сферы и регионы являются наиболее активными.

---

## **2. Количество работодателей**
- **Запрос:** Подсчет уникальных работодателей.
- **Результат:** В базе содержится **23501 работодатель**.
- **Вывод:**
  - Разнообразие работодателей - от крупных корпораций до локальных бизнесов - демонстрирует широкий спектр предложений.
  - Интересно изучить концентрацию вакансий среди лидирующих работодателей, таких как "Яндекс" или "Сбер".
  - Также важно исследовать, какие индустрии наиболее популярны среди этих работодателей.

---

## **3. Количество регионов**
- **Запрос:** Подсчет уникальных регионов.
- **Результат:** В базе указано **1362 региона**.
- **Вывод:**
  - Географическое распределение впечатляет: данные охватывают регионы по всей стране.
  - Можно исследовать различия в активности на рынке труда между крупными городами (Москва, Санкт-Петербург) и менее крупными регионами.
  - Это предоставляет возможности для анализа региональных диспропорций в доступности работы.

---

## **4. Количество сфер деятельности**
- **Запрос:** Подсчет уникальных сфер деятельности.
- **Результат:** В базе содержится **294 уникальные сферы деятельности**.
- **Вывод:**
  - Наличие почти 300 сфер деятельности демонстрирует разнообразие профессиональных направлений.
  - Интересно оценить, какие сферы занимают лидирующие позиции по количеству вакансий, например, IT, финансы или маркетинг.
  - Важно изучить, какие сферы растут быстрее остальных, чтобы определить перспективные направления.

---

## **Общий вывод**

1. **Масштаб данных:**
   - Большое количество вакансий, работодателей и регионов указывает на то, что база является комплексной и универсальной.
   - Она охватывает широкий спектр профессиональных направлений, позволяя составить полную картину рынка труда.

2. **Инсайты и тенденции:**
   - Вероятно, основная концентрация вакансий сосредоточена в крупных городах (Москва, Санкт-Петербург).
   - IT и смежные сферы, скорее всего, лидируют по количеству вакансий и уровню зарплат.
   - Необходимо подтвердить, какие индустрии являются драйверами роста, а какие остаются на периферии.

3. **Потенциальные исследования:**
   - Анализ динамики спроса на специалистов в разных регионах и сферах деятельности.
   - Оценка востребованных навыков и должностей.
   - Исследование влияния гибких графиков и удалённой работы на рынок труда.

4. **Вопросы для дальнейшего анализа:**
   - Какой процент вакансий приходится на столичные города по сравнению с регионами?
   - Какие работодатели являются лидерами по количеству предложений?
   - Как распределяются зарплатные вилки по регионам и профессиональным направлениям?

**Вывод:** База данных предоставляет мощный инструмент для анализа рынка труда, выявления тенденций и составления прогноза. Её комплексность позволяет проводить как локальные, так и глобальные исследования, что делает её ценным ресурсом для понимания текущего состояния и динамики рынка.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [15]:
# SQL-запрос для подсчета количества вакансий по регионам:
# Из таблицы public.vacancies объединяем данные с таблицей public.areas по полю area_id
# Считаем количество вакансий для каждого региона, сортируем по убыванию количества и берем топ-5
query = f"""
    SELECT 
        a.name,              -- Название региона
        COUNT(v.id) AS count -- Количество вакансий в регионе
    FROM public.vacancies v
    JOIN public.areas a
        ON v.area_id = a.id  -- Соединение таблиц по area_id
    GROUP BY a.name          -- Группировка по названию региона
    ORDER BY count DESC      -- Сортировка по количеству вакансий в убывающем порядке
    LIMIT 5                  -- Ограничение на 5 записей
"""

In [16]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [17]:
# SQL-запрос для подсчета количества вакансий, где указана зарплата:
# Проверяем, что хотя бы одно из полей salary_from или salary_to не равно NULL
query = f"""
    SELECT
        COUNT(*)
    FROM public.vacancies
    WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
"""

In [18]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [19]:
query = f"""
SELECT 
    ROUND(AVG(salary_from)) AS avg_salary_from, -- Среднее значение нижней границы зарплат
    ROUND(AVG(salary_to)) AS avg_salary_to     -- Среднее значение верхней границы зарплат
FROM public.vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL; -- Учитываем только вакансии с указанными границами зарплат
"""

In [20]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [21]:
# SQL-запрос для подсчета количества вакансий по графику работы и занятости:
# Группируем данные по полям schedule и employment, сортируем по убыванию количества и берем топ-2
query = f"""
    SELECT
        schedule,           -- График работы (например, полный день, сменный)
        employment,         -- Тип занятости (например, полная занятость, частичная)
        COUNT(*) AS count   -- Количество вакансий для каждой комбинации
    FROM public.vacancies
    GROUP BY schedule, employment -- Группировка по графику и типу занятости
    ORDER BY count DESC           -- Сортировка по количеству вакансий в убывающем порядке
    LIMIT 2                       -- Ограничение на 2 записи
"""

In [22]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

schedule        employment  count
0       Полный день  Полная занятость  35367
1  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [23]:
# SQL-запрос для подсчета количества вакансий по опыту работы:
# Проверяем, что поле experience не равно NULL, группируем по уровню опыта и сортируем по возрастанию количества
query = f"""
    SELECT
        experience,        -- Уровень опыта (например, "Нет опыта", "От 1 года")
        COUNT(*) AS count  -- Количество вакансий для каждого уровня опыта
    FROM public.vacancies
    WHERE experience IS NOT NULL -- Учитываем только строки, где поле experience заполнено
    GROUP BY experience          -- Группируем по уровню опыта
    ORDER BY count ASC           -- Сортируем по количеству вакансий в порядке возрастания
"""

In [24]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# Предварительный анализ данных

## 1. Количество вакансий по регионам
- **Наибольшее количество вакансий** сосредоточено в следующих городах:
  - Москва: **5333 вакансии**.
  - Санкт-Петербург: **2851 вакансия**.
  - Минск: **2112 вакансий**.
- Значительная активность также отмечена в:
  - Новосибирске: **2006 вакансий**.
  - Алматы: **1892 вакансии**.
- **Вывод**: Крупные города остаются основными центрами занятости, предлагая наибольшее количество вакансий.

---

## 2. Заполнение данных о зарплатах
- Всего **24073 вакансии** содержат хотя бы одно значение в полях зарплатной вилки (`salary_from` или `salary_to`).
- **Вывод**: Почти половина вакансий указывает зарплаты, но значительная часть информации о зарплатах отсутствует, что может усложнять анализ рынка.

---

## 3. Средние значения зарплатных вилок
- Среднее значение:
  - **Нижней границы (`salary_from`)**: **71,065 рублей**.
  - **Верхней границы (`salary_to`)**: **110,537 рублей**.
- **Вывод**: Разрыв между нижней и верхней границей подтверждает высокую вариативность зарплат на рынке труда.

---

## 4. Количество вакансий по графику работы и типу занятости
- **Полный рабочий день и полная занятость** лидируют — **35,367 вакансий**.
- **Удалённая работа с полной занятостью** занимает второе место — **7,802 вакансии**.
- **Вывод**: Удалённый формат труда становится популярнее, особенно в условиях полной занятости.

---

## 5. Требуемый опыт работы
- Большинство вакансий рассчитаны на кандидатов с опытом:
  - **От 1 года до 3 лет**: **26,152 вакансии**.
  - **От 3 до 6 лет**: **14,511 вакансий**.
- Для новичков и кандидатов без опыта — **7,197 вакансий**.
- **Вывод**: Несмотря на доминирование вакансий для опытных специалистов, рынок предлагает достаточное количество возможностей для начинающих.

---

# Дальнейшие шаги анализа

1. **Анализ зарплат по регионам и профессиям.** 
   - Изучить, как варьируются зарплаты в зависимости от географии и отрасли.

2. **Динамика спроса.**
   - Рассмотреть тенденции на рынке труда в зависимости от города.

3. **Глубокий анализ графиков работы.**
   - Оценить популярность удалённой работы в разных регионах и сферах.

4. **Анализ вакансий без указания зарплат.**
   - Проверить, есть ли корреляции между отсутствием зарплатных данных и типом занятости или отраслью.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [25]:
# SQL-запрос для подсчета количества вакансий у работодателей:
# Соединяем таблицы public.vacancies и public.employers по идентификатору работодателя
# Подсчитываем количество вакансий для каждого работодателя, сортируем по убыванию и берем топ-5
query = f"""
    SELECT
        e.name AS employer_name, -- Название работодателя
        COUNT(v.id) AS vacancy_count -- Количество вакансий
    FROM public.vacancies v
    JOIN public.employers e
        ON e.id = v.employer_id -- Соединение по идентификатору работодателя
    GROUP BY e.name             -- Группируем по названию работодателя
    ORDER BY vacancy_count DESC -- Сортируем по количеству вакансий в убывающем порядке
    LIMIT 5                     -- Ограничиваем результат до 5 записей
"""

In [26]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

employer_name  vacancy_count
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [27]:
# SQL-запрос для анализа регионов:
# Находим регионы, где есть работодатели, но нет вакансий
# Считаем количество работодателей и вакансий по регионам
query = f"""
    SELECT
        a.name AS reg_name,          -- Название региона
        COUNT(e.name) AS emp_count, -- Количество работодателей в регионе
        COUNT(v.id) AS vac_count    -- Количество вакансий в регионе (будет 0)
    FROM public.areas a
    JOIN public.employers e
        ON a.id = e.area            -- Соединение по региону работодателя
    LEFT JOIN public.vacancies v
        ON v.area_id = a.id         -- Левое соединение с вакансиями (возможно NULL)
    GROUP BY a.name                 -- Группировка по названию региона
    HAVING COUNT(v.id) = 0          -- Оставляем только регионы без вакансий
    ORDER BY emp_count DESC         -- Сортируем по количеству работодателей в убывающем порядке
"""

In [28]:

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

reg_name  emp_count  vac_count
0                                       Россия        410          0
1                                    Казахстан        207          0
2                           Московская область         75          0
3                           Краснодарский край         19          0
4                                     Беларусь         18          0
..                                         ...        ...        ...
181                                    Удобная          1          0
182                Майна (Ульяновская область)          1          0
183                            Верхнетуломский          1          0
184                                    Жаркент          1          0
185  Михайловка (Уфимский район, Башкортостан)          1          0

[186 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [29]:
# SQL-запрос для подсчета количества уникальных регионов, где работодатели размещают вакансии:
# Соединяем таблицы public.employers и public.vacancies по идентификатору работодателя
# Подсчитываем количество уникальных регионов, где у каждого работодателя есть вакансии
query = """
    SELECT
        e.name AS employer_name,      -- Название работодателя
        COUNT(DISTINCT v.area_id) AS unique_regions_count -- Количество уникальных регионов с вакансиями
    FROM public.employers e
    JOIN public.vacancies v
        ON e.id = v.employer_id       -- Соединение по идентификатору работодателя
    GROUP BY e.name                   -- Группировка по имени работодателя
    ORDER BY unique_regions_count DESC -- Сортировка по убыванию количества уникальных регионов
"""

In [30]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

employer_name  unique_regions_count
0                      Яндекс                   181
1                  Ростелеком                   152
2                  Спецремонт                   116
3      Поляков Денис Иванович                    88
4                   ООО ЕФИН                     71
...                       ...                   ...
14761                  UniSol                     1
14762            UNISTORY LLC                     1
14763                   UNIT6                     1
14764     United Distribution                     1
14765              UNITEL LLC                     1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [31]:
# SQL-запрос для подсчета количества работодателей, у которых нет привязки к индустриям:
# Выполняем левое соединение таблиц public.employers и public.employers_industries
# Считаем количество записей, где industry_id имеет значение NULL
query = """
    SELECT
        COUNT(e.id) AS employers_without_industry -- Подсчет количества работодателей без привязки к индустрии
    FROM public.employers e
    LEFT JOIN public.employers_industries i
        ON e.id = i.employer_id -- Соединение по идентификатору работодателя
    WHERE i.industry_id IS NULL -- Оставляем только записи без привязки к индустриям
"""

In [32]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

employers_without_industry
0                        8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [33]:
# SQL-запрос для получения работодателей, связанных ровно с четырьмя индустриями:
# Соединяем таблицы public.employers и public.employers_industries, группируем по имени работодателя
# Выбираем только тех, у кого ровно 4 записи в таблице employers_industries
query = """
    SELECT
        e.name AS employer_name,      -- Название работодателя
        COUNT(i.industry_id) AS industry_count -- Количество индустрий, связанных с работодателем
    FROM public.employers e
    JOIN public.employers_industries i
        ON e.id = i.employer_id       -- Соединение по идентификатору работодателя
    GROUP BY e.name                   -- Группировка по имени работодателя
    HAVING COUNT(i.industry_id) = 4   -- Условие: ровно 4 индустрии
    ORDER BY employer_name ASC        -- Сортировка по имени работодателя в алфавитном порядке
"""

In [34]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

employer_name  industry_count
0        101 Интернет               4
1            21vek.by               4
2                2ГИС               4
3                  2К               4
4         4 пикселя +               4
...               ...             ...
1131            ЮРИОН               4
1132  ЮТИП Технологии               4
1133           ЯКласс               4
1134            ЯрНео               4
1135    Яшин&Партнёры               4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [35]:
# SQL-запрос для подсчета количества работодателей в индустрии "Разработка программного обеспечения":
# Соединяем таблицы employers, employers_industries и industries
query = """
    SELECT
        COUNT(e.id) AS employer_count -- Подсчет количества работодателей
    FROM public.employers e
    JOIN public.employers_industries i
        ON e.id = i.employer_id      -- Соединение по идентификатору работодателя
    JOIN public.industries d
        ON i.industry_id = d.id      -- Соединение по идентификатору индустрии
    WHERE d.name = 'Разработка программного обеспечения' -- Фильтр по названию индустрии
"""

In [36]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

employer_count
0            3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [37]:
import requests
from bs4 import BeautifulSoup

# Получение списка городов из Википедии
def get_city_list():
    url = "https://ru.wikipedia.org/wiki/Города-миллионеры_России"
    headers = {"User-Agent": "Mozilla/5.0"}
    response = requests.get(url, headers=headers)
    soup = BeautifulSoup(response.text, "html.parser")
    table = soup.find("table", {"class": "standard sortable"})

    cities = []
    if table:
        rows = table.find_all("tr")[1:]  # Пропускаем заголовок
        for row in rows:
            columns = row.find_all("td")
            if columns:
                city_name = columns[1].get_text(strip=True)
                cities.append(city_name)
    return cities

In [38]:
# Формирование SQL-запроса
cities = get_city_list()
city_list_str = ", ".join([f"'{city}'" for city in cities])

query = f"""
WITH city_list AS (
    SELECT UNNEST(ARRAY[{city_list_str}]) AS city
)
SELECT
    a.name AS city,
    COUNT(v.id) AS vacancy_count
FROM public.vacancies v
JOIN public.employers e
    ON v.employer_id = e.id
JOIN public.areas a
    ON v.area_id = a.id
WHERE e.name = 'Яндекс'
  AND a.name IN (SELECT city FROM city_list)
GROUP BY a.name

UNION ALL

SELECT
    'Total' AS city,
    COUNT(v.id) AS vacancy_count
FROM public.vacancies v
JOIN public.employers e
    ON v.employer_id = e.id
JOIN public.areas a
    ON v.area_id = a.id
WHERE e.name = 'Яндекс'
  AND a.name IN (SELECT city FROM city_list)

ORDER BY vacancy_count ASC;
"""

In [39]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

city  vacancy_count
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7            Самара             26
8               Уфа             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

***

# Подробный анализ данных вакансий

## 1. Топ работодателей по количеству вакансий

### Результаты:
1. **Яндекс** — 1933 вакансии.
2. **Ростелеком** — 491 вакансия.
3. **Тинькофф** — 444 вакансии.
4. **СБЕР** — 428 вакансий.
5. **Газпром нефть** — 331 вакансия.

### Выводы:
- **Яндекс** уверенно лидирует, демонстрируя стратегию масштабного найма и усиления своих позиций в ключевых направлениях (IT, аналитика, маркетинг).
- Рост компаний **Ростелеком** и **Тинькофф** может свидетельствовать о масштабных цифровых трансформациях.
- Компании, такие как **СБЕР** и **Газпром нефть**, адаптируются к новым рыночным вызовам, создавая вакансии в области финтеха и инновационной энергетики.

---

## 2. Регионы с работодателями, но без активных вакансий

### Результаты:
1. **Россия** — 410 работодателей.
2. **Казахстан** — 207 работодателей.
3. **Московская область** — 75 работодателей.

### Выводы:
- Большое количество работодателей без активных вакансий может говорить о низкой активности рынка труда в этих регионах.
- Для **России** это может означать, что работодатели сосредоточены на поддержании текущего штата.
- **Казахстан**, на втором месте, вероятно, находится на этапе экономического перехода или сталкивается с трудностями в поиске кадров.

---

## 3. Уникальные регионы размещения вакансий работодателей

### Результаты:
1. **Яндекс** — 181 регион.
2. **Ростелеком** — 152 региона.
3. **Спецмонтаж** — 116 регионов.

### Выводы:
- Широкий географический охват **Яндекса** подтверждает его стратегию охвата как крупных городов, так и удаленных регионов, включая удаленный найм.
- **Ростелеком** демонстрирует высокую активность благодаря телекоммуникационным проектам.
- Компании, такие как **Спецмонтаж**, ориентируются на строительство и инфраструктуру в регионах.

---

## 4. Работодатели без указанной сферы деятельности

### Результат:
- 8419 работодателей не указали сферу деятельности.

### Выводы:
- Большое количество работодателей без привязки к индустрии говорит о проблемах с классификацией данных, что затрудняет их анализ.
- Вероятно, это компании малого и среднего бизнеса или индивидуальные предприниматели, которые не считают нужным указывать свою специализацию.

---

## 5. Компании с четырьмя сферами деятельности

### Результат:
- Третьей компанией по алфавиту является **2ГИС**.

### Выводы:
- Компании с несколькими сферами деятельности, как **2ГИС**, демонстрируют устойчивость, гибкость и способность диверсифицировать свои бизнес-направления.
- Такие компании часто работают на стыке IT, рекламы и аналитики.

---

## 6. Разработка программного обеспечения

### Результат:
- 3553 работодателя указали эту сферу деятельности.

### Выводы:
- **Разработка ПО** остается ключевой отраслью, обеспечивающей значительный спрос на специалистов.
- Это подчеркивает рост цифровой трансформации во всех секторах экономики.

---

## 7. Анализ вакансий Яндекса в городах-миллионниках

### Результаты:
1. **Москва** — 54 вакансии.
2. **Санкт-Петербург** — 39 вакансий.
3. **Екатеринбург** — 36 вакансий.
4. **Новосибирск** — 35 вакансий.
5. **Казань** — 25 вакансий.

### Выводы:
- **Москва** и **Санкт-Петербург** продолжают лидировать как центры экономической активности и инноваций.
- Растущая активность в городах, таких как **Новосибирск** и **Казань**, указывает на их значимость в развитии регионального бизнеса и инноваций.

---

## Общие выводы

### 1. Рынок вакансий:
- Крупные компании, такие как **Яндекс**, **Ростелеком** и **Тинькофф**, активно формируют рынок труда, предлагая широкий спектр вакансий.
- Географический охват работодателей демонстрирует потенциал развития удаленной работы и региональной экспансии.

### 2. Сферы деятельности:
- Разработка программного обеспечения сохраняет ключевую роль на рынке, поддерживая высокий спрос на IT-специалистов.
- Компании с несколькими сферами деятельности демонстрируют гибкость и адаптацию к изменениям рынка.

### 3. Регионы:
- Москва и Санкт-Петербург остаются ключевыми центрами, но внимание работодателей к региональным городам растет.
- Большое число работодателей без вакансий подчеркивает необходимость улучшения классификации данных и стимулирования активности на рынке труда.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# SQL-запрос для подсчета количества вакансий, связанных с данными:
# Фильтруем вакансии, название которых содержит 'data' или 'данн'
query = """
    SELECT
        COUNT(name) AS vacancy_count -- Подсчет количества подходящих вакансий
    FROM public.vacancies
    WHERE name ILIKE '%data%' OR name ILIKE '%данн%' -- Фильтр по названиям вакансий
"""


In [41]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

vacancy_count
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [42]:
# SQL-запрос для подсчета количества вакансий по Data Science и Machine Learning на уровне Junior или стажировки
query = """
    SELECT
        COUNT(name) AS vacancy_count -- Подсчет количества подходящих вакансий
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND
        (
            name ILIKE '%junior%' OR
            experience ILIKE '%Нет опыта%' OR
            employment ILIKE '%Стажировка%'
        )
"""

In [43]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)# результат запроса

vacancy_count
0             51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [44]:
# SQL-запрос для подсчета количества вакансий по Data Science и Machine Learning с ключевыми навыками SQL или PostgreSQL
query = """
    SELECT
        COUNT(name) AS vacancy_count -- Подсчет количества подходящих вакансий
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND
        (
            key_skills ILIKE '%SQL%' OR
            key_skills ILIKE '%postgres%'
        )
"""

In [45]:

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

vacancy_count
0            229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [46]:
# SQL-запрос для подсчета количества вакансий по Data Science и Machine Learning с ключевым навыком Python
query = """
    SELECT
        COUNT(name) AS vacancy_count -- Подсчет количества подходящих вакансий
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND
        key_skills ILIKE '%python%' -- Проверяем, что Python указан в ключевых навыках
"""

In [47]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

vacancy_count
0            357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [48]:
# SQL-запрос для подсчета среднего количества ключевых навыков в вакансиях, связанных с Data Science и Machine Learning
query = """
    SELECT
        ROUND(
            AVG(
                LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1
            ), 2
        ) AS avg_key_skills -- Среднее количество ключевых навыков, округленное до 2 знаков
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND key_skills IS NOT NULL -- Исключаем вакансии без указанных ключевых навыков
        AND key_skills <> '';      -- Исключаем вакансии с пустыми ключевыми навыками
"""

In [49]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем результат
display(df)

avg_key_skills
0            6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [50]:
# SQL-запрос для расчета средней зарплаты по опыту для вакансий Data Science и Machine Learning
query = """
    SELECT
        experience, -- Уровень опыта (например, "Нет опыта", "От 1 года")
        ROUND(AVG(
            CASE
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
                ELSE COALESCE(salary_from, salary_to) -- Используем либо salary_from, либо salary_to
            END
        ), 2) AS avg_salary -- Средняя зарплата, округленная до 2 знаков
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND (salary_from IS NOT NULL OR salary_to IS NOT NULL) -- Исключаем вакансии без указанных зарплат
    GROUP BY experience -- Группируем результаты по опыту
    ORDER BY experience -- Сортируем по уровню опыта
"""

In [51]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем результат
display(df)

experience  avg_salary
0           Нет опыта    74642.86
1  От 1 года до 3 лет   139674.75
2       От 3 до 6 лет   243114.67

***

# Анализ вакансий, связанных с данными и Data Science

## 1. Количество вакансий, имеющих отношение к данным
- **Результат:** 1771 вакансия.
- **Вывод:** Количество вакансий, в названии которых встречаются слова "data" или "данн", указывает на высокий интерес работодателей к специалистам, работающим с данными.

---

## 2. Вакансии для начинающего Data Scientist уровня Junior
- **Результат:** 51 вакансия.
- **Вывод:** Начинающим специалистам в области Data Science доступно ограниченное количество вакансий. Это подчеркивает конкуренцию на начальном уровне и необходимость наличия базовых навыков и стажировок.

---

## 3. Вакансии для DS с ключевыми навыками SQL или PostgreSQL
- **Результат:** 229 вакансий.
- **Вывод:** SQL и PostgreSQL остаются ключевыми навыками для специалистов Data Science. Это подчеркивает важность базовой технической подготовки.

---

## 4. Популярность Python в требованиях к DS
- **Результат:** 357 вакансий.
- **Вывод:** Python является одним из наиболее востребованных навыков для Data Science. Это подтверждает его лидерство в области анализа данных и машинного обучения.

---

## 5. Среднее количество ключевых навыков в вакансиях DS
- **Результат:** 6.41 навыка.
- **Вывод:** Работодатели ожидают широкий спектр навыков от специалистов DS, что указывает на сложность и междисциплинарный характер этой профессии.

---

## 6. Средняя зарплата для DS в зависимости от опыта
- **Результаты:**
  - **Нет опыта:** 74,642 руб.
  - **От 1 до 3 лет:** 139,674 руб.
  - **От 3 до 6 лет:** 243,114 руб.
- **Вывод:** Зарплаты в Data Science сильно зависят от опыта. Работодатели готовы инвестировать в квалифицированных специалистов, что делает данную сферу перспективной для развития.

# Общий вывод по проекту

In [52]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

In [53]:
# SQL-запрос для расчета средней зарплаты по опыту для вакансий Data Science и Machine Learning
query = """
-- SQL-запрос для анализа среднего уровня зарплат Data Scientists по регионам и уровням опыта
SELECT
    a.name AS region, -- Название региона
    v.experience,     -- Уровень опыта
    ROUND(
        AVG(
            CASE
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
                ELSE COALESCE(salary_from, salary_to) -- Используем либо salary_from, либо salary_to
            END
        ), 2
    ) AS avg_salary -- Средняя зарплата, округленная до 2 знаков
FROM public.vacancies v
JOIN public.areas a
    ON v.area_id = a.id -- Соединение вакансий с регионами
WHERE 
    (salary_from IS NOT NULL OR salary_to IS NOT NULL) -- Учитываем только вакансии с указанной зарплатой
    AND a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 
                   'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 
                   'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград') -- Ограничиваем анализ указанными регионами
    AND (
        v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%' OR
        v.name ILIKE '%исследователь данных%' OR
        v.name ILIKE '%machine learning%' OR
        (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%машинн%обучен%'
    ) -- Учитываем только вакансии для Data Scientist
GROUP BY a.name, v.experience -- Группируем данные по регионам и уровням опыта
ORDER BY region, experience; -- Сортируем по региону и уровню опыта
"""

In [54]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем результат
display(df)

region          experience  avg_salary
0           Воронеж         Более 6 лет    90000.00
1      Екатеринбург           Нет опыта    37000.00
2      Екатеринбург  От 1 года до 3 лет   167500.00
3            Казань  От 1 года до 3 лет    50000.00
4            Казань       От 3 до 6 лет   217750.00
5            Москва         Более 6 лет   200000.00
6            Москва           Нет опыта    72500.00
7            Москва  От 1 года до 3 лет   169178.57
8            Москва       От 3 до 6 лет   274230.77
9       Новосибирск           Нет опыта    80000.00
10      Новосибирск       От 3 до 6 лет   209500.00
11            Пермь  От 1 года до 3 лет    75000.00
12            Пермь       От 3 до 6 лет   200000.00
13   Ростов-на-Дону  От 1 года до 3 лет    50000.00
14   Ростов-на-Дону       От 3 до 6 лет   337500.00
15  Санкт-Петербург           Нет опыта    87666.67
16  Санкт-Петербург  От 1 года до 3 лет   190115.33
17  Санкт-Петербург       От 3 до 6 лет   271666.67
18              Уфа  От 1 года до 3 лет    90000.00

### 📊 Дополнительное исследование: Средняя зарплата Data Scientists по регионам и уровням опыта

#### **1. Результаты**
- **Москва**:
  - **Нет опыта**: 72,500 рублей.
  - **От 1 года до 3 лет**: 169,178.57 рублей.
  - **От 3 до 6 лет**: 274,230.77 рублей.
  - **Более 6 лет**: 200,000 рублей.

- **Санкт-Петербург**:
  - **Нет опыта**: 87,666.67 рублей.
  - **От 1 года до 3 лет**: 190,115.33 рублей.
  - **От 3 до 6 лет**: 271,666.67 рублей.

- **Новосибирск**:
  - **Нет опыта**: 80,000 рублей.
  - **От 3 до 6 лет**: 209,500 рублей.

- **Екатеринбург**:
  - **Нет опыта**: 37,000 рублей.
  - **От 1 года до 3 лет**: 167,500 рублей.

- **Казань**:
  - **От 1 года до 3 лет**: 50,000 рублей.
  - **От 3 до 6 лет**: 217,750 рублей.

- **Ростов-на-Дону**:
  - **От 1 года до 3 лет**: 50,000 рублей.
  - **От 3 до 6 лет**: 337,500 рублей (аномально высокая).

- **Пермь**:
  - **От 1 года до 3 лет**: 75,000 рублей.
  - **От 3 до 6 лет**: 200,000 рублей.

#### **2. Выводы**
1. **Москва лидирует по зарплатам:**
   - Средние зарплаты для всех уровней опыта в Москве самые высокие. Например, зарплата для специалистов с опытом от 3 до 6 лет составляет 274,230.77 рублей, что выше, чем в других регионах.

2. **Санкт-Петербург и Новосибирск:**
   - Санкт-Петербург близок к Москве по зарплатам, особенно для опытных специалистов.
   - Новосибирск выделяется относительно высокими зарплатами (209,500 рублей) для специалистов с опытом от 3 до 6 лет.

3. **Низкие зарплаты в Казани и Екатеринбурге:**
   - В Казани и Екатеринбурге зарплаты для специалистов без опыта значительно ниже: 50,000 и 37,000 рублей соответственно.

4. **Аномалия в Ростове-на-Дону:**
   - Зарплата в Ростове-на-Дону для специалистов с опытом от 3 до 6 лет составляет 337,500 рублей, что значительно выше среднего. Требуется дополнительный анализ.

#### **3. Тенденции**
1. **Опыт определяет зарплату:**
   - Зарплаты значительно увеличиваются с опытом, особенно в Москве и Санкт-Петербурге.
   
2. **Региональные различия:**
   - Зарплаты в центральных регионах (Москва, Санкт-Петербург) значительно выше, чем в остальных.

3. **Важность крупных центров:**
   - Новосибирск и Екатеринбург выделяются среди региональных центров относительно высокими зарплатами для опытных специалистов.

#### **4. Рекомендации для дальнейшего анализа**
1. **Анализ аномалий:**
   - Исследовать причину аномально высокой зарплаты в Ростове-на-Дону для специалистов с опытом от 3 до 6 лет.

2. **Динамика зарплат:**
   - Сравнить эти данные с предыдущими годами, чтобы выявить тенденции роста зарплат.

3. **Распределение вакансий:**
   - Проанализировать количество вакансий в каждом регионе, чтобы понять влияние конкуренции на зарплаты.

4. **Сравнение с другими профессиями:**
   - Сравнить зарплаты Data Scientists с зарплатами разработчиков и аналитиков в тех же регионах.

---

In [55]:
# SQL-запрос для расчета средней зарплаты по опыту для вакансий Data Science и Machine Learning
query = """
-- SQL-запрос для сравнения средних зарплат Data Scientists с другими IT-специалистами
SELECT
    CASE
        WHEN v.name ILIKE '%data scientist%' OR
             v.name ILIKE '%data science%' OR
             v.name ILIKE '%исследователь данных%' OR
             v.name ILIKE '%machine learning%' OR
             (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
             v.name ILIKE '%машинн%обучен%' THEN 'Data Scientist'
        WHEN v.name ILIKE '%разработчик%' OR
             v.name ILIKE '%developer%' THEN 'Разработчик'
        WHEN v.name ILIKE '%аналитик%' OR
             v.name ILIKE '%analyst%' THEN 'Аналитик'
        ELSE 'Другая IT-профессия'
    END AS profession, -- Классификация профессий
    ROUND(
        AVG(
            CASE
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
                ELSE COALESCE(salary_from, salary_to) -- Используем либо salary_from, либо salary_to
            END
        ), 2
    ) AS avg_salary -- Средняя зарплата, округленная до 2 знаков
FROM public.vacancies v
WHERE 
    (v.name ILIKE '%data scientist%' OR
     v.name ILIKE '%data science%' OR
     v.name ILIKE '%исследователь данных%' OR
     v.name ILIKE '%machine learning%' OR
     (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
     v.name ILIKE '%машинн%обучен%' OR
     v.name ILIKE '%разработчик%' OR
     v.name ILIKE '%developer%' OR
     v.name ILIKE '%аналитик%' OR
     v.name ILIKE '%analyst%') -- Учитываем только IT-профессии
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL) -- Исключаем вакансии без указанных зарплат
GROUP BY profession -- Группируем по классификации профессий
ORDER BY avg_salary DESC; -- Сортируем по средней зарплате в убывающем порядке
"""

In [56]:
# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем результат
display(df)

profession  avg_salary
0  Data Scientist   193860.61
1     Разработчик   140942.90
2        Аналитик    90439.47

In [57]:
connection.close()

### 📊 Вывод по дополнительному исследованию: сравнение средних зарплат Data Scientists с другими IT-специалистами

#### **1. Результаты**
- **Data Scientist**: средняя зарплата — **193,860.61 рублей**.
- **Разработчик**: средняя зарплата — **140,942.90 рублей**.
- **Аналитик**: средняя зарплата — **90,439.47 рублей**.

#### **2. Выводы**
1. **Data Scientists лидируют по уровню зарплат:**
   - Data Scientists получают наибольшую среднюю зарплату среди всех анализируемых IT-специалистов. Это связано с высокой сложностью работы, необходимостью глубоких знаний в области математики, программирования и анализа данных.

2. **Разработчики занимают второе место:**
   - Разработчики получают в среднем на 27% меньше, чем Data Scientists, но их зарплаты остаются достаточно высокими. Это объясняется стабильным спросом на разработчиков в широком спектре индустрий.

3. **Аналитики с самыми низкими зарплатами:**
   - Аналитики данных получают на 53% меньше, чем Data Scientists. Это может быть связано с более низкими требованиями к техническим навыкам по сравнению с Data Scientists и разработчиками.

#### **3. Тенденции и прогнозы**
1. **Рост спроса на Data Scientists:**
   - Ожидается, что зарплаты Data Scientists будут продолжать расти благодаря растущему спросу на анализ данных и машинное обучение в бизнесе.

2. **Позиция разработчиков:**
   - Несмотря на более низкий уровень зарплат по сравнению с Data Scientists, разработчики остаются одной из самых востребованных категорий специалистов в IT.

3. **Возможности для аналитиков:**
   - Аналитики могут увеличивать свою стоимость на рынке, если будут развивать технические навыки, такие как Python, SQL, и переходить в роли, связанные с Data Science.

#### **4. Рекомендации**
1. **Развитие компетенций:**
   - Специалистам, стремящимся увеличить свой доход, следует развивать навыки, востребованные в Data Science и разработке, такие как Python, машинное обучение и работа с большими данными.

2. **Анализ по регионам:**
   - Провести исследование, как зарплаты Data Scientists, разработчиков и аналитиков варьируются в зависимости от региона, чтобы выявить географические тренды.

3. **Динамика изменений:**
   - Сравнить эти данные с историческими, чтобы выявить рост или падение зарплат в каждой категории.

---